File: RMG.Core/Composition/EventStatePattern.fs
namespace RMG.CoreF.Composition

open RMG.CoreF

[<Sealed>]
type EventStatePattern
    private
    (
        duration: Duration,
        timeline: EventStateTimelineMap,
        patternTimeline: EventStatePattern WithEventState Timeline,
        flatTimeline: EventStateTimelineMap
    ) =
    member this.Duration = duration
    member this.Timeline = timeline

    member this.PatternTimeline = patternTimeline

    member this.FlatTimeline = flatTimeline

    static member internal ofTimelinesUnsafe
        (duration: Duration)
        (timeline: EventStateTimelineMap)
        (patternTimeline: EventStatePattern WithEventState Timeline)
        (flatTimeline: EventStateTimelineMap)
        : EventStatePattern =
        EventStatePattern(duration, timeline, patternTimeline, flatTimeline)

module EventStatePattern =
    let ofTimelines (duration: Duration) (timeline: EventStateTimelineMap) (patternTimeline: EventStatePattern WithEventState Timeline) : EventStatePattern =
        let flatTimeline =
            seq {
                yield struct (EventState.empty, timeline) |> Timeline.itemOfSingle
                yield! patternTimeline |> Timeline.map (fun struct (eventState, pattern) -> struct (eventState, pattern.FlatTimeline))
            }
            |> Timeline.ofSeq
            |> EventStateTimelineMap.concatCombined duration

        EventStatePattern.ofTimelinesUnsafe duration timeline patternTimeline flatTimeline

File: Rmg.Core/Composition/EventStatePattern.cs
using Rmg.Core.Events;

namespace Rmg.Core.Composition;

public sealed class EventStatePattern<T>
    where T : notnull
{
    public static EventStatePattern<T> Create(
        double duration,
        EventStateTimelineMap<T> timelineMap,
        EventTimeline<WithStateMap<EventStatePattern<T>>> patternTimeline
    )
    {
        ArgumentOutOfRangeException.ThrowIfNegative(duration);
        
        var mergedStateTimelineMap = patternTimeline
            .MapValues(x => x.Value.MergeStateMap(x.StateMap))
            .Unwrap()
            .MergeWith(timelineMap);
        return new EventStatePattern<T>(duration, timelineMap, patternTimeline, mergedStateTimelineMap);
    }

    public double Duration { get; }

    public EventStateTimelineMap<T> TimelineMap { get; }

    public EventTimeline<WithStateMap<EventStatePattern<T>>> PatternTimeline { get; }

    public EventStateTimelineMap<T> MergedTimelineMap { get; }

    private EventStatePattern(
        double duration,
        EventStateTimelineMap<T> timelineMap,
        EventTimeline<WithStateMap<EventStatePattern<T>>> patternTimeline,
        EventStateTimelineMap<T> mergedTimelineMap
    )
    {
        ArgumentOutOfRangeException.ThrowIfNegative(duration);
        
        Duration = duration;
        TimelineMap = timelineMap;
        PatternTimeline = patternTimeline;
        MergedTimelineMap = mergedTimelineMap;
    }

    public EventStateTimelineMap<T> MergeStateMap(StateMap stateMap)
    {
        return MergedTimelineMap.MergeStateMap(stateMap);
    }
}

public static class EventStatePattern
{
    public static EventStatePattern<T> Create<T>(
        double duration,
        EventStateTimelineMap<T> timelineMap,
        EventTimeline<WithStateMap<EventStatePattern<T>>> patternTimeline
    ) where T : notnull
    {
        return EventStatePattern<T>.Create(duration, timelineMap, patternTimeline);
    }
}
